File: src/fpga/kernels/gemm.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from ..transport import FpgaExecutor, GemmTileI16Request, GemmTileShape


@dataclass(slots=True)
class MatrixI16:
    rows: int
    cols: int
    values: list[int]

    def row(self, index: int) -> list[int]:
        start = index * self.cols
        end = start + self.cols
        return self.values[start:end]


@dataclass(slots=True)
class MatrixI64:
    rows: int
    cols: int
    values: list[int]

    def get(self, row: int, col: int) -> int:
        return self.values[row * self.cols + col]


@dataclass(slots=True)
class GemmComparison:
    software: MatrixI64
    rtl: MatrixI64
    matched: bool
    notes: list[str]


def software_gemm(lhs: MatrixI16, rhs: MatrixI16) -> MatrixI64:
    return software_gemm_with_accumulator(lhs, rhs, None)


def software_gemm_with_accumulator(
    lhs: MatrixI16,
    rhs: MatrixI16,
    accumulator: MatrixI64 | None,
) -> MatrixI64:
    if lhs.cols != rhs.rows:
        raise ValueError(
            f"incompatible GEMM dimensions: lhs {lhs.rows}x{lhs.cols}, rhs {rhs.rows}x{rhs.cols}"
        )

    if accumulator is not None:
        if accumulator.rows != lhs.rows or accumulator.cols != rhs.cols:
            raise ValueError(
                "accumulator shape mismatch: "
                f"expected {lhs.rows}x{rhs.cols}, got {accumulator.rows}x{accumulator.cols}"
            )

    values: list[int] = []
    for row in range(lhs.rows):
        for col in range(rhs.cols):
            total = 0 if accumulator is None else accumulator.get(row, col)
            for inner in range(lhs.cols):
                lhs_value = int(lhs.values[row * lhs.cols + inner])
                rhs_value = int(rhs.values[inner * rhs.cols + col])
                total += lhs_value * rhs_value
            values.append(total)

    return MatrixI64(rows=lhs.rows, cols=rhs.cols, values=values)


def simulate_gemm_tile(
    executor: FpgaExecutor,
    output_dir: Path,
    audio_path: str,
    lhs: MatrixI16,
    rhs: MatrixI16,
) -> GemmComparison:
    return simulate_gemm_tile_with_accumulator(
        executor=executor,
        output_dir=output_dir,
        audio_path=audio_path,
        lhs=lhs,
        rhs=rhs,
        accumulator=None,
    )


def simulate_gemm_tile_with_accumulator(
    executor: FpgaExecutor,
    output_dir: Path,
    audio_path: str,
    lhs: MatrixI16,
    rhs: MatrixI16,
    accumulator: MatrixI64 | None,
) -> GemmComparison:
    if lhs.cols != rhs.rows:
        raise ValueError(
            f"incompatible GEMM dimensions: lhs {lhs.rows}x{lhs.cols}, rhs {rhs.rows}x{rhs.cols}"
        )

    software = software_gemm_with_accumulator(lhs, rhs, accumulator)

    response = executor.execute_gemm_tile(
        GemmTileI16Request(
            audio_path=audio_path,
            shape=GemmTileShape(rows=lhs.rows, cols=rhs.cols, inner=lhs.cols),
            lhs_tile=list(lhs.values),
            rhs_tile=list(rhs.values),
            accumulator_input=(
                [0 for _ in range(lhs.rows * rhs.cols)]
                if accumulator is None
                else list(accumulator.values)
            ),
            expected_output=list(software.values),
        ),
        output_dir,
    )

    rtl = MatrixI64(rows=lhs.rows, cols=rhs.cols, values=list(response.rtl_output))
    return GemmComparison(
        software=software,
        rtl=rtl,
        matched=bool(response.matched),
        notes=list(response.notes),
    )

File: src/fpga/kernels/gemm.rs
use anyhow::{Result, bail};

use crate::fpga::kernels::dot::{simulate_dot_product, software_dot_product};
use crate::fpga::transport::FpgaExecutor;

#[derive(Debug, Clone)]
pub struct MatrixI16 {
    pub rows: usize,
    pub cols: usize,
    pub values: Vec<i16>,
}

impl MatrixI16 {
    pub fn new(rows: usize, cols: usize, values: Vec<i16>) -> Self {
        Self { rows, cols, values }
    }

    pub fn row(&self, index: usize) -> &[i16] {
        let start = index * self.cols;
        let end = start + self.cols;
        &self.values[start..end]
    }

    pub fn get(&self, row: usize, col: usize) -> i64 {
        i64::from(self.values[row * self.cols + col])
    }
}

#[derive(Debug, Clone)]
pub struct MatrixI64 {
    pub rows: usize,
    pub cols: usize,
    pub values: Vec<i64>,
}

impl MatrixI64 {
    pub fn new(rows: usize, cols: usize, values: Vec<i64>) -> Self {
        Self { rows, cols, values }
    }

    pub fn get(&self, row: usize, col: usize) -> i64 {
        self.values[row * self.cols + col]
    }
}

#[derive(Debug, Clone)]
pub struct GemmComparison {
    pub software: MatrixI64,
    pub rtl: MatrixI64,
    pub matched: bool,
    pub notes: Vec<String>,
}

pub fn software_gemm(lhs: &MatrixI16, rhs: &MatrixI16) -> Result<MatrixI64> {
    if lhs.cols != rhs.rows {
        bail!(
            "incompatible GEMM dimensions: lhs {}x{}, rhs {}x{}",
            lhs.rows,
            lhs.cols,
            rhs.rows,
            rhs.cols
        );
    }

    let mut values = Vec::with_capacity(lhs.rows * rhs.cols);
    for row in 0..lhs.rows {
        for col in 0..rhs.cols {
            let rhs_column = (0..rhs.rows)
                .map(|rhs_row| rhs.values[rhs_row * rhs.cols + col])
                .collect::<Vec<_>>();
            values.push(software_dot_product(lhs.row(row), &rhs_column));
        }
    }

    Ok(MatrixI64::new(lhs.rows, rhs.cols, values))
}

pub fn simulate_gemm_via_dot_products(
    executor: &dyn FpgaExecutor,
    output_dir: &std::path::Path,
    audio_path: &str,
    lhs: &MatrixI16,
    rhs: &MatrixI16,
) -> Result<GemmComparison> {
    if lhs.cols != rhs.rows {
        bail!(
            "incompatible GEMM dimensions: lhs {}x{}, rhs {}x{}",
            lhs.rows,
            lhs.cols,
            rhs.rows,
            rhs.cols
        );
    }
    if lhs.cols != 8 {
        bail!("current dot-product RTL path expects inner dimension of exactly 8");
    }

    let software = software_gemm(lhs, rhs)?;
    let mut rtl_values = Vec::with_capacity(lhs.rows * rhs.cols);
    let mut notes = Vec::new();
    let mut matched = true;

    for row in 0..lhs.rows {
        for col in 0..rhs.cols {
            let rhs_column = (0..rhs.rows)
                .map(|rhs_row| rhs.values[rhs_row * rhs.cols + col])
                .collect::<Vec<_>>();
            let result =
                simulate_dot_product(executor, output_dir, audio_path, lhs.row(row), &rhs_column)?;
            matched &= result.matched;
            rtl_values.push(result.rtl_result);
            notes.extend(result.notes);
        }
    }

    Ok(GemmComparison {
        software,
        rtl: MatrixI64::new(lhs.rows, rhs.cols, rtl_values),
        matched,
        notes,
    })
}
